src/utils.js

```diff
@@ -0,0 +1,62 @@
+
+/**
+ * Variant of Promise.race(iterable)
+ *
+ * Given a list of Promises, return a Promise which resolves
+ * with the index of the first Promise in the list which resolves
+ * AND doesnt explicitly return false,
+ * ensuring that all Promises before it have been settled.
+ *
+ * E.g.:
+ *
+ *   When the structure of promiseList is:
+ *
+ *   //   [<rejects in 1s>, <resolves in 1s>, <resolved>]
+ *   // => Resolves with #2 after 1s
+ *
+ *   //   [<resolved>, <resolves in 1s>, <resolved>]
+ *   // => Resolves with #1 immediately
+ *
+ *   //   [<rejects in 1s>, <rejected>, <resolved>]
+ *   // => Resolves with #3 after 1s
+ *
+ *   //   [<rejected>, <resolves in 1s>, <resolved>]
+ *   // => Resolves with #2 after 1s
+ *
+ *   //   [<resolves in 2s>, <resolves in 1s>, <resolved>]
+ *   // => Resolves with #1 after 2s
+ *
+ * @param {Array} promiseList
+ * @returns {Promise}
+ */
+export const PromiseOrderedFirst = function(promiseList) {
+  return new Promise((resolve, reject) => {
+    const promiseToStateMap = [];
+    const compute = () => {
+      for (const { state, res, index } of promiseToStateMap) {
+        if (state === "pending") {
+          return;
+        } else if (state === "resolved" && res !== false) {
+          resolve({res, index});
+        }
+      }
+      reject();
+    };
+    promiseList.forEach((promiseLike, index) => {
+      promiseToStateMap[index] = { state: "pending" };
+      promiseLike
+        .then(res => {
+          promiseToStateMap[index] = { state: "resolved", res, index };
+        })
+        .catch(res => {
+          promiseToStateMap[index] = { state: "rejected", res, index };
+        })
+        .then(compute, compute);
```

Comment:
we want to compute twice? or am I misunderstanding this

Reply:
No, this just means we invoke `compute` on resolve or on reject of the promise.

Signature: `.then(<onFulfilled>, <onRejected>)`

This is a safer alternative/fallback to `.finally(<onFulfilled or rejected>)` because `.finally` is only supported starting in Chrome 63.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/finally

Reply:
ohhh right thanks for the explanation! Did not realize that about finally.